File: 2022/06/14_reverse_integer/yongsu.c
#include <stdio.h>
#include <string.h>

// 문제가 하라는 대로 품

long long my_power(int num, int exp)
{
    int i;
    long long ans = 1;
    for (i = 0; i < exp; i++)
    {
        ans *= num;
    }
    return ans;
}

int reverse(int x){
    char buffer[15];
    int i;
    int len;
    int negative = 0;

    long long x_longlong = (long long) x;

    long long ans = 0;

    for (i = 0; i < 15; i++) buffer[i] = 0;

    if (x_longlong < 0) {
        x_longlong = -x_longlong;
        negative = 1;
    }

    sprintf(buffer, "%llu", x_longlong);

    len = strlen(buffer);

    for (i = len - 1; i >= 0; i--) ans += ((int)buffer[i] - 48) * my_power(10, i);

    if (negative) ans = -ans;

    if (ans >= my_power(2, 31) - 1 || ans <= -(my_power(2, 31) - 1)) ans = 0;

    return (int) ans;
}

// Runtime: 0 ms, faster than 100.00% of C online submissions for Reverse Integer.
// Memory Usage: 5.8 MB, less than 11.46% of C online submissions for Reverse Integer.

File: 2022/06/23_generate_parentheses/yongsu.py
from typing import List
import unittest

# 너무피곤해서 파이썬으로 함

# Runtime: 74 ms, faster than 13.69% of Python3 online submissions for Generate Parentheses.
# Memory Usage: 14.3 MB, less than 12.19% of Python3 online submissions for Generate Parentheses.

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []            

        def dfs(element: str, open: int, close: int):
            if open == close == n:
                ans.append(element)
                return

            if open < n:
                next_element = element + '('
                dfs(next_element, open + 1, close)
            if close < open:
                next_element = element + ')'
                dfs(next_element, open, close + 1)
        
        dfs("", 0, 0)
        return ans


class SolutionTests(unittest.TestCase):
    def setUp(self) -> None:
        self.sol = Solution()

    def test_1(self):
        answer = ["((()))","(()())","(())()","()(())","()()()"]
        result = self.sol.generateParenthesis(3)
        self.assertEqual(result, answer)


if __name__ == "__main__":
    unittest.main()